File: src/main/java/weld/WeldError.java
package weld;

/**
 * Wrapper for the native weld error object.
 *
 * Note that this object must be closed after usage.
 */
public class WeldError extends WeldManaged {

  public WeldError() {
    super(WeldJNI.weld_error_new());
  }

  @Override
  protected void doClose() {
    WeldJNI.weld_error_free(this.handle);
  }

  public int getCode() {
    checkAccess();
    return WeldJNI.weld_error_code(this.handle);
  }

  public String getMessage() {
    checkAccess();
    return WeldJNI.weld_error_message(this.handle);
  }
}

File: src/main/scala/weld/WeldError.scala
package weld

class WeldError extends WeldManaged(WeldJNI.weld_error_new()) {
  override protected def doClose(): Unit = {
    WeldJNI.weld_error_free(handle)
  }

  override protected def cleaner = new WeldError.Cleaner(handle)

  def code: Int = {
    checkAccess()
    WeldJNI.weld_error_code(handle)
  }

  def message: String = {
    checkAccess()
    WeldJNI.weld_error_message(handle)
  }
}

object WeldError {
  private[weld] class Cleaner(handle: Long) extends Runnable {
    override def run(): Unit = WeldJNI.weld_error_free(handle)
  }
}

class WeldException(val code: Int, message: String) extends RuntimeException(message) {
  def this(message: String) = this(-1, message)
  def this(error: WeldError, code: Option[String] = None) =
    this(error.code, error.message + code.map("\n" + _).getOrElse(""))
}
